### ISSUE TITLE
System.IO.IOException: The response ended prematurely

### ISSUE THREAD

**Issue by chsakell** [dotnet/runtime#57030]
### Description

After migrating from `.NET Core 2.2` to `.NET Core 3.1` we randomly get the following error:

```csharp
System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The response ended prematurely.
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
 ```

* This is not something we can reproduce on a low traffic environment but it does happen on our high traffic production website.
* Obviously this relates to [47612](https://github.com/dotnet/runtime/issues/47612) and [31528](https://github.com/dotnet/runtime/issues/31528) - I don't think it's a coincidence that the issue happened after migrating to .NET Core 3.1 or .NET 5.0
* The client uses an `HttpClient` injected (`services.AddHttpClient`)
* Nothing changed on the server that returns the response

### Additional info

This is how HttpClient is configured:

```csharp
services.AddScoped<NativeApiHttpClientHandler>();
services.AddHttpClient<INativeApiClient, NativeApiClient>()
    .ConfigurePrimaryHttpMessageHandler(() => new NativeApiHttpClientHandler(useHttpProxy))
    .ConfigureHttpClient(httpClient =>
    {
        if (httpClientTimeout != null) httpClient.Timeout = httpClientTimeout.Value;
    });
```

```csharp
public NativeApiHttpClientHandler(bool? useProxy)
{
    UseProxy = useProxy ?? true;
    UseCookies = false;
    AllowAutoRedirect = false;
    AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
}
```

I am not sure if it helps but here's how the response is deserialized:

```csharp
using (var s = await  response.Content.ReadAsStreamAsync())
using (var sr = new StreamReader(s))
using (var reader = new JsonTextReader(sr))
{
    var serializer = new JsonSerializer
    {
        ContractResolver = SerializationUtils.NativeApiContractResolver,
        NullValueHandling = NullValueHandling.Ignore
    };

    serializer.Error += (se, ev) =>
    {
        ev.ErrorContext.Handled = true;
    };

    result = serializer.Deserialize<NativeApiBase<T>>(reader) ?? new NativeApiBase<T>();
}
```

**Comment by danmoseley** [comment]
Is it possible to say whether it happens on 5.0 (or even 6.0 preview 6)?

**Comment by chsakell** [comment]
Difficult to say as we don't plan to migrate on 5.0 soon, we will migrate on 6.0 when it's stable.

**Comment by stephentoub** [comment]
What is NativeApiHttpClientHandler?

**Comment by chsakell** [comment]
[Configure the HttpMessageHandler](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-5.0#configure-the-httpmessagehandler)
Just configuring the inner `HttpMessageHandler`

**Comment by stephentoub** [comment]
Right,  but what type is that?

**Comment by chsakell** [comment]
I though I posted the entire class above, but posted just a method.. Here's the class.

```csharp
public class NativeApiHttpClientHandler : HttpClientHandler
    {
        public NativeApiHttpClientHandler(bool? useProxy)
        {
            UseProxy = useProxy ?? true;
            UseCookies = false;
            AllowAutoRedirect = false;
            AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
        }
    }
```

**Comment by chsakell** [comment]
The problem was solved by setting the following at startup:
```csharp
AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);
```
Reference [here](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler?view=netcore-3.1).

The question is, that if we do want to use the `SocketsHttpHandler`, is there any equivalent for setting properties such as `AutomaticDecompression` ? As we did using the `ConfigurePrimaryHttpMessageHandler` during registration?

**Comment by ManickaP** [comment]
@chsakell your workaround swapped managed implementation for the native one. Just FYI this option is not available in 5.0 anymore (WinHttpHandler is still avaialble but only as out-of-band nuget package).

In our experience, this error is in 99% caused by the server really closing the connection. What do we know about the server here?

Either way, to make this actionable for us, we would need a simple repro showing the problem. Is this something you could do?

**Comment by karelz** [comment]
Not actionable without additional information. Closing.
If more info becomes available, feel free to reopen.
